README.md
# ChokeOnSmoke
Mod for Factorio which adds a furnace that let's you burn fuel to create pollution

## Research "Choke on smoke" technology
![Technology](https://thehorscht.github.io/choke-on-smoke-technology.png)

## Craft a fuel burner furnace
![Recipe](https://thehorscht.github.io/choke-on-smoke-recipe.png)
## Put fuel in the bottom and the top
![Recipe](https://thehorscht.github.io/choke-on-smoke-furnace.png)
## Lean back and watch the pollution build up

changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.1.2
Date: 12. 10. 2018
  Changes:
    - Visual change only: Reduced smoke frequency to reduce processing power needed.
  Bugfixes:
    - Made furnace not-rotatable to fix smoke coming from wrong position.
---------------------------------------------------------------------------------------------------
Version: 0.1.1
Date: 12. 10. 2018
  Bugfixes:
    - Corrected description.


prototypes/fuel.lua
local _ = require "name-uniquifier"

-- Define a new fluid type that never gets used except as an icon for ALT mode,
-- to show that "pollution" is being "crafted"
local pollution = table.deepcopy(data.raw.fluid.steam)
pollution.name = _("pollution")
pollution.icon = "__choke-on-smoke__/graphics/icons/fluid/pollution.png"
pollution.icon_size = 32
data:extend{pollution}

-- Converts a string like "225MJ" to a number which represents the megajoules
-- 2MJ wooden chest nuclear fuel 1.21GJ
function parse_joule_string_to_mjs(joule_str)
  local amount, unit = string.match(joule_str, "^(.+)([kKMGTPEZY]J)$")
  amount = tonumber(amount)
  if amount == nil or unit == nil then
    return nil
  end
  local powers_of_ten = {
    kJ = -3, KJ = -3,
    MJ = 0,
    GJ = 3,
    TJ = 6,
    PJ = 9,
    EJ = 12,
    ZJ = 15,
    YJ = 18,
  }
  amount = amount * 1 * 10^powers_of_ten[unit]
  return amount
end

-- Make everything thats fuel burnable in our special furnace
for k,v in pairs(data.raw.item) do
  if v.fuel_category == "chemical" then
    log(v.name ..": fuel_value parsed = ".. parse_joule_string_to_mjs(v.fuel_value))
    log(v.name ..": energy_required = ".. (1 * parse_joule_string_to_mjs(v.fuel_value)))
    data:extend{
      {
        type = "recipe",
        name = _(v.name.."-to-smoke"),
        category = _("fuel"),
        energy_required = 1 * parse_joule_string_to_mjs(v.fuel_value),
        emissions_multiplier = 1,
        hide_from_stats = true,
        hidden = true,
        ingredients = {{v.name, 1}},
        results = {{ type="fluid", name=_("pollution"), amount=0}}
      }
    }
  end
end

prototypes/furnace.lua
local _ = require "name-uniquifier"

local tint = {r = 1, g = 0.7, b = 0.7, a = 1}

data:extend{
  {
    type = "furnace",
    name = _("furnace"),
    icon = "__base__/graphics/icons/steel-furnace.png",
    icon_size = 32,
    flags = {"placeable-neutral", "placeable-player", "player-creation", "not-rotatable"},
    minable = {mining_time = 1, result = _("furnace")},
    max_health = 300,
    corpse = "medium-remnants",
    vehicle_impact_sound =  { filename = "__base__/sound/car-metal-impact.ogg", volume = 0.65 },
    working_sound =
    {
      sound = { filename = "__base__/sound/furnace.ogg" }
    },
    resistances =
    {
      {
        type = "fire",
        percent = 100
      }
    },
    collision_box = {{-1.2, -1.2}, {1.2, 1.2}},
    selection_box = {{-1.3, -1.5}, {1.3, 1.5}},
    crafting_categories = {_("fuel")},
    result_inventory_size = 0,
    energy_usage = "250kW",
    crafting_speed = 2,
    source_inventory_size = 1,
    energy_source =
    {
      type = "burner",
      fuel_category = "chemical",
      effectivity = 1.00,
      emissions = 2.0 * settings.startup[_("multiplier")].value,
      fuel_inventory_size = 1,
      smoke =
      {
        {
          name = "smoke",
          frequency = 20,
          position = {1.05, -1.7},
          starting_vertical_speed = 0.07,
          starting_frame_deviation = 60
        }
      }
    },
    animation =
    {
      layers =
      {
        {
          filename = "__base__/graphics/entity/steel-furnace/steel-furnace.png",
          priority = "high",
          width = 205,
          height = 87,
          frame_count = 1,
          shift = util.by_pixel(90.5, 3.5),
          apply_runtime_tint = true,
          tint = tint,
          scale = 1.5,
          hr_version =
          {
            filename = "__base__/graphics/entity/steel-furnace/hr-steel-furnace.png",
            priority = "high",
            width = 171,
            height = 174,
            frame_count = 1,
            shift = util.by_pixel(-1.25, 5),
            apply_runtime_tint = true,
            tint = tint,
            scale = 0.75
          }
        },
        {
          filename = "__base__/graphics/entity/steel-furnace/steel-furnace-shadow.png",
          priority = "high",
          width = 139,
          height = 43,
          frame_count = 1,
          draw_as_shadow = true,
          shift = util.by_pixel(60.5, 11.5),
          scale = 1.5,
          hr_version =
          {
            filename = "__base__/graphics/entity/steel-furnace/hr-steel-furnace-shadow.png",
            priority = "high",
            width = 277,
            height = 85,
            frame_count = 1,
            draw_as_shadow = true,
            shift = util.by_pixel(60, 15),
            scale = 0.75
          }
        }
      }
    },
    working_visualisations =
    {
      {
        north_position = {0.0, 0.0},
        east_position = {0.0, 0.0},
        south_position = {0.0, 0.0},
        west_position = {0.0, 0.0},
        animation =
        {
          filename = "__base__/graphics/entity/steel-furnace/steel-furnace-fire.png",
          priority = "high",
          line_length = 8,
          width = 29,
          height = 40,
          frame_count = 48,
          axially_symmetrical = false,
          direction_count = 1,
          shift = util.by_pixel(-0.5, 6),
          scale = 1.5,
          hr_version =
          {
            filename = "__base__/graphics/entity/steel-furnace/hr-steel-furnace-fire.png",
            priority = "high",
            line_length = 8,
            width = 57,
            height = 81,
            frame_count = 48,
            axially_symmetrical = false,
            direction_count = 1,
            shift = util.by_pixel(-0.75, 5.75),
            scale = 0.75
          }
        },
        light = {intensity = 1, size = 1, color = {r = 1.0, g = 1.0, b = 1.0}}
      },
      {
        north_position = {0.0, 0.0},
        east_position = {0.0, 0.0},
        south_position = {0.0, 0.0},
        west_position = {0.0, 0.0},
        effect = "flicker", -- changes alpha based on energy source light intensity
        animation =
        {
          filename = "__base__/graphics/entity/steel-furnace/steel-furnace-glow.png",
          priority = "high",
          width = 60,
          height = 43,
          frame_count = 1,
          shift = {0.03125, 0.640625},
          blend_mode = "additive",
          scale = 1.5
        }
      },
      {
        north_position = {0.0, 0.0},
        east_position = {0.0, 0.0},
        south_position = {0.0, 0.0},
        west_position = {0.0, 0.0},
        effect = "flicker", -- changes alpha based on energy source light intensity
        animation =
        {
          filename = "__base__/graphics/entity/steel-furnace/steel-furnace-working.png",
          priority = "high",
          line_length = 8,
          width = 64,
          height = 75,
          frame_count = 1,
          axially_symmetrical = false,
          direction_count = 1,
          shift = util.by_pixel(0, -4.5),
          scale = 1.5,
          blend_mode = "additive",
          hr_version =
          {
            filename = "__base__/graphics/entity/steel-furnace/hr-steel-furnace-working.png",
            priority = "high",
            line_length = 8,
            width = 130,
            height = 149,
            frame_count = 1,
            axially_symmetrical = false,
            direction_count = 1,
            shift = util.by_pixel(0, -4.25),
            blend_mode = "additive",
            scale = 0.75
          }
        }
      }
    },
    fast_replaceable_group = "furnace",
  }
}

-- Create an item that let's us place our furnace
data:extend{
  {
      type = "item",
      name = _("furnace"),
      icons = {
        {
          icon = "__base__/graphics/icons/steel-furnace.png",
          icon_size = 32,
          tint = tint,
        }
      },
      flags = {"goes-to-quickbar"},
      subgroup = "smelting-machine",
      order = "b[steel-furnace]",
      place_result = _("furnace"),
      stack_size = 50
  },
  {
      name = _("fuel"),
      type = "recipe-category"
  },
}